contracts/MarginCalculator.sol

```diff
@@ -2,11 +2,280 @@
+     *
+     * Formula: net = ( long cash value * long Amount ) - (short cash value * short amount)
+     *
+     * @return net value
+     */
+    function _calculateExpiredPutNetValue(
+        int256 _shortAmount,
+        int256 _longAmount,
+        int256 _shortCashValue,
+        int256 _longCashValue
+    ) internal pure returns (int256) {
+        return (_longCashValue.mul(_longAmount)).sub(_shortCashValue.mul(_shortAmount));
+    }
+
+    /**
+     * @dev calculate expired net call spread value.
+     *                     ( long cash value * long Amount ) - (short cash value * short amount)
+     *  Formula: net =   -------------------------------------------------------------------------
+     *                                               Underlying price
+     * @return net value
+     */
+    function _calculateExpiredCallNetValue(
+        int256 _shortAmount,
+        int256 _longAmount,
+        int256 _shortCashValue,
+        int256 _longCashValue,
+        int256 _underlyingPriceInt
+    ) internal pure returns (int256) {
+        return (_longCashValue.mul(_longAmount)).sub(_shortCashValue.mul(_shortAmount)).div(_underlyingPriceInt);
+    }
+
+    /**
+     * @dev internal function that ensure each asset type & amout array have the same length, and length <= 1;
+     * @param _vault the vault to check.
+     */
+    function _checkAssetCount(Vault memory _vault) internal pure {
+        // For the currect version, check lengths of short, long, ollateral <= 1.
+        require(_vault.shortOtokens.length <= 1, "MarginCalculator: Too many short otokens in the vault.");
+        require(_vault.longOtokens.length <= 1, "MarginCalculator: Too many long otokens in the vault.");
+        require(_vault.collateralAssets.length <= 1, "MarginCalculator: Too many collateral assets in the vault.");
+
+        require(
+            _vault.shortOtokens.length == _vault.shortAmounts.length,
+            "MarginCalculator: Short asset and amount mismatch."
+        );
+        require(
+            _vault.longOtokens.length == _vault.longAmounts.length,
+            "MarginCalculator: Long asset and amount mismatch."
+        );
+        require(
+            _vault.collateralAssets.length == _vault.collateralAmounts.length,
+            "MarginCalculator: Collateral asset and amount mismatch."
+        );
+    }
+
+    /**
+     * @dev internal function that check the long asset is valid for the short asset
+     * @param _vault the vault to check.
+     */
+    function _checkLongAsset(Vault memory _vault) internal view {
+        if (_vault.longOtokens.length == 0 || _vault.shortOtokens.length == 0) return;
+
+        OtokenInterface long = OtokenInterface(_vault.longOtokens[0]);
+        OtokenInterface short = OtokenInterface(_vault.shortOtokens[0]);
+
+        require(long.expiryTimestamp() == short.expiryTimestamp(), "MarginCalculator: Short and Long expiry mismatch.");
+        require(
```

Comment:
why not just check that batchId for both is the same?

Reply:
remove the 3 require statements for the batchId one

Reply:
Damn you're smart

Reply:
Great suggestion! I think It's cleaner. I'll adopt it for now. 

But just want to mention that using hash is gonna be more expensive, and the revert message will be less useful. Happy to have more discussion if anyone has other preference.

Reply:
Also, in the future version, if we add calendar spread, then we need to check things separately.